docs(manifest): Improvements to `scope` page

dipikabh opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This PR is part of improving the web/manifest docs:

- Adds "Syntax", "Values", and "Description"
- Covers notes and examples from the spec
- Adds prose to the examples

### Motivation

To ensure all sections have sufficient explanation, all caveats from spec are covered, and the manifest pages follow a similar template

### Additional details

Spec links:
- https://w3c.github.io/manifest/#scope-member
- https://w3c.github.io/manifest/#nav-scope

### Related issues and pull requests

Tracking issue: https://github.com/mdn/mdn/issues/560

Also:
Fixes https://github.com/mdn/content/issues/32948

dipikabh committed:
update scope page

dipikabh committed:
update syntax

dipikabh committed:
edit value desc

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
 
-The `scope` member is a string that defines the navigation scope of this web application's application context. It restricts what web pages can be viewed while the manifest is applied. If the user navigates outside the scope, it reverts to a normal web page inside a browser tab or window.
+The `scope` manifest member is used to define the navigation scope of your web application.
+It restricts the web pages that can be navigated while the manifest is applied.
```
At this point even worth stating that it is a URL, and that it restricts the scope to paths below the indicated path (not just "in" the indicated path).

Maybe 

```suggestion
It restricts the web pages that can be navigated while the manifest is applied to be a sub-page of the specified path.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
+A URL is considered to be "within scope" if its path begins with the URL path defined in `scope`.
+For example, if the `scope` is set to `/app/`, then the URLs `/app/`, `/app/page.html`, and `/app/dashboard/index.html` will all be considered within scope, while `/` or `/page.html` will not be considered within the scope of the application's context.
+
+If a user navigates to a web page that is not within the scope of your web app's context, browsers may display a prominent UI element, which clearly shows the URL or at least its origin, including whether it is served over a secure connection.
+This UI element will typically be different from what users see when navigating within the app's scope.
+This behavior aims to make users aware that they have navigated away from your web app.
```
> It is not clear to me what the visual difference would be in and out of the app. What you say is that it will be visually different. My assumption is that the visual element only belongs to web apps and so you'll always have that element, and then lose it when you navigate out. Can we have a screenshot of both cases?

This is comment I put in the intro. It would be better to word as "when within the app a UI xxx is applied. When outside the app YYY occurs, so it is obvious that you're outside the application context.". Screenshot of at least one implementation showing the UI element also pretty important.

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
-Finally, the following example limits navigation to a subdirectory of the current site:
+### Limiting navigation to a specific part in your web app
+
+If you have a web app with multiple sections, but you want to limit the navigation to one section, you can define the `scope` as:
```
We don't limit navigation
```suggestion
If you have a web app with multiple sections, but you want to limit the application scope to one section, you can define the `scope` as:
```

dipikabh replied:
Updated to:

> If you have a website with multiple sections, but you want your web app to focus on a specific section, you can define the `scope` as:

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
 {{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}
 
-The `scope` member is a string that defines the navigation scope of this web application's application context. It restricts what web pages can be viewed while the manifest is applied. If the user navigates outside the scope, it reverts to a normal web page inside a browser tab or window.
+The `scope` manifest member is used to define the navigation scope of your web application.
```
Maybe
```suggestion
The `scope` manifest member is used to define the top level URL paths that contains your web application.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
 {{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}
 
-The `scope` member is a string that defines the navigation scope of this web application's application context. It restricts what web pages can be viewed while the manifest is applied. If the user navigates outside the scope, it reverts to a normal web page inside a browser tab or window.
+The `scope` manifest member is used to define the navigation scope of your web application.
+It restricts the web pages that can be navigated while the manifest is applied.
```
1. At this point even worth stating that it is a URL, and that it restricts the scope to paths below the indicated path (not just "in" the indicated path).
2. I don't really like the term "navigation scope" either. Because again, it kind of implies a restriction to navigation.
3. According to your documentation below this is wrong. It isn't a restriction since you can navigate anywhere. All it does is make clear that you are in or out of the app.
4. It is not clear to me what the visual difference would be in and out of the app. What you say is that it will be visually different. My assumption is that the visual element only belongs to web apps and so you'll always have that element, and then _lose it_ when you navigate out. Can we have a screenshot of both cases? 

Assuming I am right, how about something like:

```suggestion
The `scope` manifest member is used to define the top level URL path that contains your web application.
Web pages that are part of the web app must be nested in some path below the specified URL, and are highlighted so that it is obvious when a user is navigating within the application.
```

dipikabh replied:
Wdyt about this:

> The scope manifest member is used to specify the top-level URL path that contains your web application, determining which pages and subdirectories are part of the web app and where the manifest is applied.
> Web pages within the defined scope are presented distinctly so that users can recognize when they are navigating within the app.

- From the spec (https://w3c.github.io/manifest/#abstract):
    "Additionally, the manifest allows a developer to "scope" a web application to a URL. This restricts the URLs to which the manifest is applied and provides a means to "deep link" into a web application from other applications."

- "Navigation scope" is a term used in the spec but I understand your concern.
       - "The manifest's scope member is a string that represents the navigation scope of this web application's application context." (https://w3c.github.io/manifest/#scope-member)
        - "The navigation scope of a manifest is the "scope" item of a processed manifest. The navigation scope restricts the set of URLs to which an application context can be navigated while the manifest is applied." (https://w3c.github.io/manifest/#dfn-navigation-scope)

hamishwillee replied:
As per https://github.com/mdn/content/pull/36137#discussion_r1805554059 the text you put in does not fix the problem. 
Nor does my text.

In essence if you are browsing a web app it looks like you're in an app. If you have a scope to your app and you navigate out of that scope the user is still inside what looks like an app. So this scope is all about showing that you're no longer in the app even though you're still in the same kind of environment.
That wasn't obvious to me, and is IMO perhaps worth spelling out.

I repeat this point (sorry) in https://github.com/mdn/content/pull/36137#discussion_r1805572580 and https://github.com/mdn/content/pull/36137#discussion_r1805583452

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
+
+When `scope` is missing or invalid, it defaults to `start_url`. For example:
+
+- If `start_url` is `https://example.com/app/index.html?user=123#home`, the navigation scope will be `https://example.com/app/`.
+- If `start_url` is `/pages/welcome.html`, the navigation scope will be `/pages/` on the same origin.
+- If `start_url` is `/pages/` (the trailing slash is important), the navigation scope will be `/pages/`.
```
```suggestion
- If `start_url` is `https://example.com/app/index.html?user=123#home`, the scope will be `https://example.com/app/`.
- If `start_url` is `/pages/welcome.html`, the scope will be `/pages/` on the same origin.
- If `start_url` is `/pages/` (the trailing slash is important), the scope will be `/pages/`.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
+For example, the target URL string `/prefix-of/resource.html` will match an app with `scope` set as `/prefix`, even though the path segment name is not an exact match.
+To avoid unexpected behavior, it's recommended to use a scope ending with a `/`.
+
+### Navigation scope
```
I am OK with "Navigation here"

dipikabh replied:
I've actually updated this to "In-scope and out-of-scope behavior"

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
````diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
 ```
 
-Finally, the following example limits navigation to a subdirectory of the current site:
+### Limiting navigation to a specific part in your web app
````
Maybe? Again, we're not limiting navigation, just scope. Also not to a specific part of our web app, but our web app to a specific part of our site.


```suggestion
### Limiting a web app to a sub path of your site
```

dipikabh replied:
Updated to:

```
### Defining a web app for a specific section of your site
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
````diff
@@ -7,32 +7,103 @@ browser-compat: html.manifest.scope
+}
 ```
 
+Users will be able to navigate to `https://example.com/store/products`, but navigating to `https://example.com/company/` will be out of scope for your web app. For off-scope URLs, browsers may display different UI elements to let users know they've navigated away from the app.
````
Same point - users will be able to navigate to wherever they like. Leaving you to reword around this. Also check for cases I might have missed commenting above.

dipikabh replied:
Updated as:

> With this setup, pages like `https://example.com/store/products` are part of your web app, but `https://example.com/company/` is out of your web app's scope. For off-scope URLs, browsers may display different UI elements to let users know they've navigated away from the app.

Also updated the first two examples

dipikabh committed:
fix feedback

dipikabh committed:
fix feedback

dipikabh commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,116 @@ browser-compat: html.manifest.scope
+
+The `scope` member defines the set of URLs that is considered part of your web app when the manifest is applied.
+
+It enables deep linking into your web app from other applications. For example, if your hiking app's scope is `https://hikingapp.com/trails/`, a weather app can directly link to `https://hikingapp.com/trails/trailA/trail-conditions`. Note that deep linking is possible even without defining a scope, but defining it allows you to control pages that are considered part of your web app.
```
Added a point about deep linking here based on the spec (https://w3c.github.io/manifest/#abstract):

"Additionally, the manifest allows a developer to "scope" a web application to a URL. This restricts the URLs to which the manifest is applied and provides a means to "deep link" into a web application from other applications."

hamishwillee replied:
Sorry, this makes no sense at all.

> It enables deep linking into your web app from other applications. 

Implies that deep linking into an app requires scope.

> For example, if your hiking app's scope is `https://hikingapp.com/trails/`, a weather app can directly link to `https://hikingapp.com/trails/trailA/trail-conditions`. 

That just looks like an URL? How is this link different to a normal URL. If it is somehow special, then I would probably remove the example and link to some definition of deep linking that covers the format.

> Note that deep linking is possible even without defining a scope, but defining it allows you to control pages that are considered part of your web app.

So now we don't need scope to deep link.

And we're repeating that scope lets us control what is part of the app.

dipikabh committed:
update nav scope

dipikabh commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,114 @@ browser-compat: html.manifest.scope
+
+### Fallback scope behavior
+
+The `scope` is invalid if `start_url` is not a subset of the `scope` URL. For example:
+
+- Valid: `scope` is `/app/` and `start_url` is `/app/home.html`.
+- Invalid: `scope` is `/app/` and `start_url` is `/index.html`.
```
Added example for `start_url` not within defined scope

dipikabh commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,114 @@ browser-compat: html.manifest.scope
+
+### Scope matching mechanism
+
+String matching for the scope URL uses a simple prefix match, not the path structure.
+For example, if the `scope` is set as `/prefix`, it will match URLs starting with `/prefix`, including `/prefix-of/index.html` and `/prefix/index.html`. Note that `/prefix-of/index.html` matches even though `prefix-of` is not an exact match with the scope `/prefix`.
+
+For this reason, it's recommended to define a scope ending with a `/`.
+Setting the `scope` as `/prefix/` ensures it will match only the pages within the `/prefix/` directory, preventing unintended matches.
```
Explained a bit more

dipikabh commented:
Thanks for the review, Hamish.

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,32 +7,114 @@ browser-compat: html.manifest.scope
+
+## Description
+
+The `scope` member defines the set of URLs that is considered part of your web app when the manifest is applied.
```
This is redundant

```suggestion
The `scope` member defines the set of URLs that is considered part of your web app.
```

hamishwillee replied:
The term "when the manifest is applied" actually means "when browsing the site as a web app" - I think the second form is much more useful to readers because it implies the difference that matters.

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,73 +36,84 @@ It restricts the web pages that can be navigated while the manifest is applied.
 
-If a user navigates to a web page that is not within the scope of your web app's context, browsers may display a prominent UI element, which clearly shows the URL or at least its origin, including whether it is served over a secure connection.
-This UI element will typically be different from what users see when navigating within the app's scope.
+When a user navigates to a web page that is not within the scope of your web app's application context, browsers may display a prominent UI element, which clearly shows the URL or at least its origin, including whether it is served over a secure connection.
+This UI element will be different from what users see when navigating within the app's scope.
```
OK, so I found this confusing because at first I assumed navigating outside of your web app meant "browsing in a normal browser".

What I think you are saying is that when you install a web app and start browsing in it you get a different UI experience than browsing in a web browser - one that looks like an app.
When you navigate out of the part of  your site that is inside your web app scope you are still browsing inside your web app "UI" - the browser doesn't magically turn into a normal browser.

So what happens is that there is some UI change that you have outlined here to show that. Right?

That is probably bleedingly obvious, but perhaps so obvious that I missed it. Might be wroth spelling out in the intro?

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,73 +36,84 @@ It restricts the web pages that can be navigated while the manifest is applied.
-This UI element will typically be different from what users see when navigating within the app's scope.
+When a user navigates to a web page that is not within the scope of your web app's application context, browsers may display a prominent UI element, which clearly shows the URL or at least its origin, including whether it is served over a secure connection.
+This UI element will be different from what users see when navigating within the app's scope.
 This behavior aims to make users aware that they have navigated away from your web app.
```
As above ^^^. You're still in what looks like a web app/doesn't look like a browser.

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,73 +36,84 @@ It restricts the web pages that can be navigated while the manifest is applied.
 > [!NOTE]
-> The existence of a `scope` doesn't prevent navigation to URLs outside of the defined scope while the manifest is applied.
-> Off-scope navigations are not blocked and not opened in a new top-level browsing context.
+> The existence of a `scope` doesn't prevent users from navigating to URLs outside of the defined scope while the manifest is applied.
+> Off-scope navigations are not blocked by browsers and are not opened in a new top-level browsing context.
```
Some redundant words there
```suggestion
> Web app users can navigate to URLs outside of the defined scope.
> Off-scope navigations are not blocked and are not opened in a new top-level browsing context.
```

I like the words, but this is back to my point https://github.com/mdn/content/pull/36137/files/647b31cdba854996e4576779b691e03c4a755e25..2bfcd2315cabfc9283896a3ce18b7d3090c92dcd#r1805572580

What this "means" is:
- If you're in a web app it displays like an app, without the browser etc.
- If you navigate outside of the scope then you are still in the web app framework - it doesn't become a browser.
  So to make it clear you're no longer browsing in the web app they need to do something else/display a UI element.
- If you are browsing the same website in a browser you will see no effect of scope or the manifest.

dipikabh committed:
fix feedback comments

dipikabh committed:
change position of the note

dipikabh commented:
Hi Hamish, thanks a lot for the review!

I've done a couple of things in this iteration:
- Clarified in the intro and in other sections that the behavior is applicable and visible with the installed app (not when using the app in a browser).
- Moved "[In-scope and out-of-scope behavior](https://pr36137.content.dev.mdn.mozit.cloud/en-US/docs/Web/Manifest/scope#in-scope_and_out-of-scope_behavior)" to make it the first subsection in "Description" since it explains the core functionality.
  - I've added a web app example and screenshots that hopefully better illustrate the in-scope and out-of-scope behaviors.
- Added a separate subsection for "[Scope's affect on deep-linked pages](https://pr36137.content.dev.mdn.mozit.cloud/en-US/docs/Web/Manifest/scope#scopes_affect_on_deep-linked_pages)" and reworded prose to improve clarity

I'm hoping these updates will address your comments. Let me know what you think.

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -7,8 +7,9 @@ browser-compat: html.manifest.scope
 
-The `scope` manifest member is used to specify the top-level URL path that contains your web application, determining which pages and subdirectories are part of the web app and where the manifest is applied.
-Web pages within the defined scope are presented distinctly so that users can recognize when they are navigating within the app.
+The `scope` manifest member is used to specify the top-level URL path that contains your web application's pages and subdirectories.
+When users install and use your web app, pages _within scope_ provide an app-like interface.
+When users navigate to pages outside the app's scope, they still experience the app-like interface, but browsers display UI elements like the URL bar to indicate the change in context.
```
Yes, this is really clear

hamishwillee replied:
Something like this?

```suggestion
The `scope` manifest member is used to specify the top-level URL path that defines the pages and subdirectories of your website that are considered to be part of your web application.

When users install and use your web app, the website is browsed within an app-like interface.
If the user browses outside of the part of the site that is within the scope, or to another completely different website, they still experience the app-like interface.
To indicate that users are browsing outside of the context of the web app, browsers display UI elements like the URL bar.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,9 +37,57 @@ Web pages within the defined scope are presented distinctly so that users can re
 ## Description
 
-The `scope` member defines the set of URLs that is considered part of your web app when the manifest is applied.
+The `scope` member defines the URLs that are part of your web app's installed experience. Browsers use `scope` to determine if a page is within your web app's {{Glossary("Application_context", "application context")}}.
```
```suggestion
The `scope` member defines the URLs that are part of your web app's installed experience.
Browsers use `scope` to determine if a page is within your web app's {{Glossary("Application_context", "application context")}}.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,9 +37,57 @@ Web pages within the defined scope are presented distinctly so that users can re
+A URL is considered to be "within scope" if its path begins with the URL path defined in `scope`.
+For example, if the `scope` is set to `/app/`, then the URLs `/app/`, `/app/page.html`, and `/app/dashboard/index.html` are all considered within scope, while `/` or `/page.html` are not.
+
+When users browse your installed web app, they experience an app-like interface.
```
```suggestion
When users open your installed web app, they experience an app-like interface.
```

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,9 +37,57 @@ Web pages within the defined scope are presented distinctly so that users can re
+For example, if the `scope` is set to `/app/`, then the URLs `/app/`, `/app/page.html`, and `/app/dashboard/index.html` are all considered within scope, while `/` or `/page.html` are not.
+
+When users browse your installed web app, they experience an app-like interface.
+For in-scope pages, browsers maintain the application context and preserve the app-like experience without browser controls.
+When users navigate to pages outside the app's scope, they still experience the app-like interface; however, in these pages, browsers display a prominent UI element like the URL bar.
+This helps users understand that they're viewing pages outside the app's defined scope.
```
```suggestion
For in-scope pages, browsers maintain the application context and preserve the app-like experience.
When users navigate to pages outside the app's scope, they still broadly experience the app-like interface; however, in these pages, browsers display additional UI element like the URL bar.
This helps users understand that they're viewing pages outside the app's defined scope.
```

dipikabh replied:
Fair! Updated

hamishwillee commented on files/en-us/web/manifest/scope/index.md:
```diff
@@ -36,9 +37,57 @@ Web pages within the defined scope are presented distinctly so that users can re
+| ----------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------- |
+| ![Trail listing page showing app-like interface without browser controls](trail-list.png) | ![Blog page showing website address and browser controls while maintaining app-like interface](blog.png) |
+
+### Scope's affect on deep-linked pages
```
I get this now! Thanks

hamishwillee reviewed (commented):
Changes look really good. I've added some very minor suggestions.

dipikabh committed:
apply minor suggestions

dipikabh commented:
Thanks, I've incorporated your suggestions.

hamishwillee reviewed (approved):
Thanks very much! Looking good.
